### PR TITLE
Task 07 userspace memory tests implementation

### PR DESCRIPTION
**Pull request by DrCaH4ec**
Done part of task 07 with the implementation of the memory allocation speed tests in the userspace.

### CODE REVIEW

File: 07_memory/user_mem/logs.rst [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+    Start adress is 0x7fffb2049340
+    Alloc time = 3231ns     Free time = 0ns
+    Iter No21
+    Start adress is 0x7fffb1f49340
+    Alloc time = 4879ns     Free time = 0ns
+    Iter No22
+    Start adress is 0x7fffb1d49340
+    Alloc time = 3527ns     Free time = 0ns
+    Iter No23
+    Segmentation fault (core dumped)
```
Comment (ekovalyov) [inline 07_memory/user_mem/logs.rst]: Is it possible to change program for exclude this segmentation fault?

Comment (DrCaH4ec) [reply 07_memory/user_mem/logs.rst]: Yes, at least we can not refer to the returned pointer.

File: 07_memory/user_mem/logs.rst [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+    Alloc time = 15ns   Free time = 0ns
+    Iter No59
+    Alloc time = 21ns   Free time = 0ns
+    Iter No60
+    Alloc time = 15ns   Free time = 0ns
+    Iter No61
+    Alloc time = 14ns   Free time = 0ns
+    Iter No62
+    Alloc time = 16ns   Free time = 0ns
+    Iter No63
```
Comment (ekovalyov) [inline 07_memory/user_mem/logs.rst]: How many bytes are allocated by Iter No63?

File: 07_memory/user_mem/user_mem.c [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+{
+	if (t1.tv_sec == t2.tv_sec)
+		return (t1.tv_nsec - t2.tv_nsec);
+
+	uint64_t tmp_nsec = (NSEC_IN_SEC + t1.tv_nsec) - t2.tv_nsec;
+
+	uint64_t tmp_sec = tmp_nsec / NSEC_IN_SEC;
+	tmp_sec += (t1.tv_sec - t2.tv_sec - 1);
+
+	return (tmp_sec * NSEC_IN_SEC) + (tmp_nsec % NSEC_IN_SEC);
```
Comment (ekovalyov) [inline 07_memory/user_mem/user_mem.c]: Is it possible to simplify the function?

File: 07_memory/user_mem/user_mem.c [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+		if (func != ALLOCA_F)
+			printf("Start adress is %p\n", t->ptr);
+		
+		printf("Alloc time = %ldns\t", t->alloc_time);
+		printf("Free time = %ldns\n", t->free_time);
+	}
+
+	printf("Aver allocating time is %ldns\t", t_aver->alloc_time / MAX_BUF_SIZE);
+	printf("Aver freeing time is %ldns\n", t_aver->free_time / MAX_BUF_SIZE);
```
Comment (ekovalyov) [inline 07_memory/user_mem/user_mem.c]: Can You implement the average value for every size.

File: 07_memory/user_mem/logs.rst (hunk 1 above)
Comment (ekovalyov) [reply 07_memory/user_mem/logs.rst]: This is not enough.

File: 07_memory/user_mem/logs.rst [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+        38      0               41              56              |       0               0               0               274877906944
+        39      0               40              56              |       0               0               0               549755813888
+        40      0               40              56              |       0               0               0               1099511627776
+        41      0               40              56              |       0               0               0               2199023255552
+        42      0               41              56              |       0               0               0               4398046511104
+        43      0               40              56              |       0               0               0               8796093022208
+        44      0               40              56              |       0               0               0               17592186044416
+        45      0               40              56              |       0               0               0               35184372088832
```
Comment (ekovalyov) [inline 07_memory/user_mem/logs.rst]: I suppose it is impossible to allocate so huge buffer on your computer. Please change your code.
We alternative with segmentation fault is better than completely wrong results.

File: 07_memory/user_mem/logs.rst (hunk 1 above)
Comment (ekovalyov) [reply 07_memory/user_mem/logs.rst]: If it is difficult to fix segfault and provide correct result please leave version with segmentation fault.

Comment (DrCaH4ec) [reply 07_memory/user_mem/logs.rst]: I've done it)

File: 07_memory/user_mem/user_mem.c (hunk 3 above)
Comment (ekovalyov) [reply 07_memory/user_mem/user_mem.c]: It seems to me this function works wrong for some arguments.

File: 07_memory/user_mem/logs.rst [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        1       0               47              397             |       0               46              149             2
+        2       0               42              397             |       0               45              149             4
+        3       0               43              397             |       0               45              149             8
+        4       0               52              397             |       0               44              149             16
+        5       0               51              397             |       0               45              149             32
+        6       0               56              397             |       0               56              149             64
+        7       0               54              397             |       0               46              149             128
+        8       0               61              397             |       0               55              149             256
+        9       0               50              397             |       0               46              149             512
```
Comment (ekovalyov) [inline 07_memory/user_mem/logs.rst]: Is minimum allocation time value 0 correct?

File: 07_memory/user_mem/user_mem.c [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+	uint64_t sec_diff;
+	uint64_t nsec_diff;
+
+	sec_diff = t1.tv_sec - t2.tv_sec;
+	nsec_diff = t1.tv_nsec - t2.tv_nsec;
+
+	if (nsec_diff < 0) {
+		nsec_diff += NSEC_IN_SEC;
+		sec_diff--;
+	}
```
Comment (ekovalyov) [inline 07_memory/user_mem/user_mem.c:31]: Is this IF block required?

Comment (DrCaH4ec) [reply 07_memory/user_mem/user_mem.c:31]: Yes, it's necessary for cases when _t1_ and _t2_ are differ for 1 or more seconds.

**Pull request merged**